### PR TITLE
time_window_partitions: add get_partition_key helper method that converts date/datetime to str partition key

### PR DESCRIPTION
**Pull request by Gw1p**
## Summary & Motivation

Create a helper method `get_partition_key` on `TimeWindowPartitionsDefinition` that allows to get a string representation of a partition key using a string, datetime or date. This can be convenient for hourly/daily/weekly/monthly partitioned assets, as the method will automatically use the relevant partition definition format.

The current implementation will error if it gets a date/datetime that is not a valid partition key for the given partition definition.

## How I Tested These Changes

Added tests in `test_partitioned_assets.py`. Also, tested in a local Dagster.

## Changelog

> Add get_partition_key helper method that can be used on hourly/daily/weekly/montly partitioned assets to get partition key using date or datetime.

### CODE REVIEW

**Review by jamiedemaria — COMMENTED** [review]
This seems like a fine change to me, but i'm not sure if there is a reason we only have allowed strings up to this point. @clairelin135 and @dpeng817 you're the most recent authors on this part of the code, do you have any additional context to add?

Would also like to see some tests for this before merging. https://github.com/dagster-io/dagster/blob/279d44aa2f325f2d2a883807da6f29bf3a2b10e1/python_modules/dagster/dagster_tests/asset_defs_tests/test_partitioned_assets.py would be a good spot to add them

**Comment by dpeng817** [comment]
I sympathize with the desire to have a convenient way of getting the partition key for an asset, but I don't love this approach. `AssetMaterialization` is a pretty core abstraction, and ultimately we expect all partition keys to be resolvable to strings. I think there's a ton of code paths that would break if partition_key was all of a sudden not necessarily a string.

Furthermore, the mapping of a datetime to a partition key depends on the actual partition definition of the asset, and in the constructor of an assetmaterialization we don't have access to that information (and I don't think we want to need access to that information). I think what you actually want to do is something like this:
```python
import dagster as dg
from datetime import datetime
MY_TIME_WINDOW_PARTITIONS_DEF = dg.DailyPartitionsDefinition(...)

my_datetime = datetime(...)

MY_TIME_WINDOW_PARTITIONS_DEF.get_partition_key(my_datetime)
```

the question is; what happens if the partition key does not align directly with the given datetime? Should we error? 

Basically, i think what we really want here are better utilities for going from a given datetime to the relevant partition key.

**Review by dpeng817 — CHANGES_REQUESTED** [review]
Left a lengthy comment, really appreciate you taking the time to write up this PR but I worry this isn't the right approach.

**Comment by Gw1p** [comment]
> I sympathize with the desire to have a convenient way of getting the partition key for an asset, but I don't love this approach. `AssetMaterialization` is a pretty core abstraction, and ultimately we expect all partition keys to be resolvable to strings. I think there's a ton of code paths that would break if partition_key was all of a sudden not necessarily a string.
> 
> Furthermore, the mapping of a datetime to a partition key depends on the actual partition definition of the asset, and in the constructor of an assetmaterialization we don't have access to that information (and I don't think we want to need access to that information). I think what you actually want to do is something like this:
> 
> ```python
> import dagster as dg
> from datetime import datetime
> MY_TIME_WINDOW_PARTITIONS_DEF = dg.DailyPartitionsDefinition(...)
> 
> my_datetime = datetime(...)
> 
> MY_TIME_WINDOW_PARTITIONS_DEF.get_partition_key(my_datetime)
> ```
> 
> the question is; what happens if the partition key does not align directly with the given datetime? Should we error?
> 
> Basically, i think what we really want here are better utilities for going from a given datetime to the relevant partition key, given a partitionsdefinition.

Makes sense. I'm happy to work on utils to derive partition from partition definition, that seems like a nicer approach.

I'm leaving for the Christmas holidays but will get on it when I return in 2 weeks.

**Comment by Gw1p** [comment]
> I sympathize with the desire to have a convenient way of getting the partition key for an asset, but I don't love this approach. `AssetMaterialization` is a pretty core abstraction, and ultimately we expect all partition keys to be resolvable to strings. I think there's a ton of code paths that would break if partition_key was all of a sudden not necessarily a string.
> 
> Furthermore, the mapping of a datetime to a partition key depends on the actual partition definition of the asset, and in the constructor of an assetmaterialization we don't have access to that information (and I don't think we want to need access to that information). I think what you actually want to do is something like this:
> 
> ```python
> import dagster as dg
> from datetime import datetime
> MY_TIME_WINDOW_PARTITIONS_DEF = dg.DailyPartitionsDefinition(...)
> 
> my_datetime = datetime(...)
> 
> MY_TIME_WINDOW_PARTITIONS_DEF.get_partition_key(my_datetime)
> ```
> 
> the question is; what happens if the partition key does not align directly with the given datetime? Should we error?
> 
> Basically, i think what we really want here are better utilities for going from a given datetime to the relevant partition key, given a partitionsdefinition.

@dpeng817 I've implemented the suggested changes. Let me know what you think about this approach!

**Review by dpeng817 — APPROVED** [review]
Okay nice. Thanks for the additional test.

**Comment by dpeng817** [comment]
@Gw1p this just needs a rebase, and I'll merge it. Thanks for your contribution!

**Comment by Gw1p** [comment]
Had a pyright failure. Should be fixed.
